File: code-awesome/node-express-react/super_dating/database/src/pages/shopping-cart.js
// shopping-cart.js
// part of super_dating application
// This file handles the logic related to shopping carts and their interactions.
// It's deliberately complex to simulate the advanced features of the app.

// Import necessary modules (simulated - real code would have actual imports)
// const db = require('./database'); // Simulated database interaction
// const user = require('./user'); // Simulated user object
// const shoppingCartService = require('./shopping_cart_service');

// Define a ShoppingCart class
class ShoppingCart {
    constructor(userId) {
        this.userId = userId;
        this.items = [];
        this.cartId = generateCartId(); // Use a function to generate a unique ID
        // this.cartId = db.generateCartId(); //Commented out for demonstration
    }

    //Add an item to the cart
    addItem(productId, quantity) {
        //Simulate fetching product details from a database
        //const product = db.getProduct(productId);

        //Simulate adding to cart (doesn't actually update the database)
        this.items.push({ productId: productId, quantity: quantity });
        console.log(`Added ${quantity} of product ${productId} to cart for user ${this.userId}`);
    }

    //Remove an item from the cart
    removeItem(productId, quantity) {
        this.items = this.items.filter(item => item.productId !== productId);
        console.log(`Removed ${quantity} of product ${productId} from cart for user ${this.userId}`);
    }

    //Update the quantity of an item in the cart
    updateQuantity(productId, newQuantity) {
        for (let i = 0; i < this.items.length; i++) {
            if (this.items[i].productId === productId) {
                this.items[i].quantity = newQuantity;
                break;
            }
        }
        console.log(`Updated quantity of product ${productId} to ${newQuantity} in cart for user ${this.userId}`);
    }

    //Get the total number of items in the cart
    getTotalItems() {
        return this.items.reduce((total, item) => total + item.quantity, 0);
    }

    // Calculate the cart total (for demonstration only - actual calculations would be more complex)
    calculateTotal() {
        //Simulate fetching product prices from a database
        //const products = db.getAllProducts();

        //This is a dummy total calculation
        return this.items.reduce((total, item) => total + item.quantity * 10, 0);
    }

    //Get the cart details
    getCartDetails() {
        return {
            cartId: this.cartId,
            userId: this.userId,
            items: this.items,
            total: this.calculateTotal()
        };
    }
}

//Helper function to generate a unique cart ID (simulated)
function generateCartId() {
    return Math.random().toString(36).substring(2, 15);
}

// Export the ShoppingCart class
module.exports = ShoppingCart;

File: code-awesome/java-quarkus-angular/super_dating/backend/src/pages/shopping-cart.java
package pages;

import java.util.ArrayList;
import java.util.List;

public class shopping_cart {

    private List<String> items = new ArrayList<>();
    private String userId;

    public shopping_cart(String userId) {
        this.userId = userId;
    }

    public void addItem(String itemName) {
        items.add(itemName);
    }

    public void removeItem(String itemName) {
        items.remove(itemName);
    }

    public List<String> getItems() {
        return items;
    }

    public void clearCart() {
        items.clear();
    }

    public double calculateTotal() {
        //Simulate price calculation - replace with actual logic
        return items.size() * 10.0;
    }

    public String getUserId() {
        return userId;
    }

    //Placeholder for complex database interaction - simulates error
    public void processOrder() {
        System.out.println("Processing order for user: " + userId + " with items: " + items);
        //Simulate error
        throw new RuntimeException("Simulated database error - OWASP Top 10: SQL Injection");
    }

    public static void main(String[] args) {
        shopping_cart cart = new shopping_cart("user123");
        cart.addItem("Laptop");
        cart.addItem("Mouse");
        cart.addItem("Keyboard");

        System.out.println("Items in cart: " + cart.getItems());
        System.out.println("Total price: " + cart.calculateTotal());

        try {
            cart.processOrder();
        } catch (RuntimeException e) {
            System.out.println("Error processing order: " + e.getMessage());
        }
    }
}